### DISCUSSION TITLE
Add adapter in Dockerfile and lunch it locally (or CI) for integration test

### DISCUSSION THREAD

**Discussion by dicrescenzoa** [awslabs/aws-lambda-web-adapter#507] [General]
Hello everyone,

We need to run the adapter locally (or in CI) in order to run integration test.

Specifically the flow that we want to cover is `apigateway invocation event -> web-adapter -> code`

Specifically we want to test how the lambda receive the invocation message and how this is going to be handled by the adapter and the http framework below.

for example with a request like [this](https://docs.aws.amazon.com/lambda/latest/dg/images-test.html#images-test-AWSbase)

We are trying to test with the go-gin example

```Dockerfile
FROM golang:1.18-alpine AS build_base
RUN apk add --no-cache git
WORKDIR /tmp/gin

COPY . .
RUN go mod download

RUN GOOS=linux CGO_ENABLED=0 go build -o bootstrap .
FROM alpine:3.9 
RUN apk add ca-certificates
COPY --from=public.ecr.aws/awsguru/aws-lambda-adapter:0.8.1 /lambda-adapter /opt/extensions/lambda-adapter
COPY --from=build_base /tmp/gin/bootstrap /app/bootstrap

ENV PORT=8000 GIN_MODE=release
EXPOSE 8000

CMD ["/app/bootstrap"]
```

We are building the image as following and running with these commands

```
docker build -t gin-web-adapter-test:latest . 

docker run -p 127.0.0.1:8080:8000 gin-web-adapter-test:latest
```

However we see that in this way we run directly the code and not the layer that should be used for converting the invocation to the http request.

Thank you for you support 👍

**Comment by dicrescenzoa** [comment]
As I see the web-adapter extension defined here
```
COPY --from=public.ecr.aws/awsguru/aws-lambda-adapter:0.8.1 /lambda-adapter /opt/extensions/lambda-adapter
```
is not executed locally since the execution of the extension is an internal lambda mechanism:

https://aws.amazon.com/blogs/compute/working-with-lambda-layers-and-extensions-in-container-images/

so how to run this in a CI/CD to run integration test? 🤔

**Comment by bnusunny** [comment]
Lambda Web Adapter is an external extension. Lambda service will automatically start it during function init. 

You can use `sam local start-api` to simulate Lambda and API Gateway. It works in CI/CD as well.

**Comment by dicrescenzoa** [comment]
HI @bnusunny thanks for the answer. 

We don't use sam in our project.  Is there a way to run the adapter in a docker container? 

Our service tests are based on [testcontainer](https://testcontainers.com/). At the moment we are testing using the [gin adapter](https://github.com/awslabs/aws-lambda-go-api-proxy/tree/master) that is deprecated and we are planning to move from zip to container and use this adapter. I'm not sure, is it correct testing http request instead that invocation in this scenario? is it the correct approach?

However, I see something similar have been attempted here (to run the layer locally)

https://github.com/awslabs/aws-lambda-web-adapter/issues/19

**Reply by bkeane** [reply]
It should be valid to run your tests against the webserver without involving the adapter. The adapter sits in your docker image as an already-tested invoke/request mapper. There is indeed a very thin integration layer test one could do for ensuring your port configuration etc is aligned for running in lambda itself. But the juice is hardly worth the squeeze.

A benefit of this adapter approach is that you can just write/test your webserver as you normally would, and rely on this thing to deal with the event-to-request mapping.

All that being said, you can for sure use the lambda rie to do all of this locally, but it's quite a lot of machinery to achieve what amounts to calling your webserver directly.

If you release yourself from trying to test lambda itself as well as the web adapter, you will have a better time.

**Comment by bnusunny** [comment]
I'm not familiar with testcontainer. Give me some time to look into it. 

You don't have to simulate the Lambda execution environment for local dev/test. Test the gin app normally gives you the best local development experience. 

For integration testing in CI/CD pipeline, it makes sense to simulate the Lambda execution environment. But some customers prefer to test against the real cloud services. They deploy the applications as real Lambda functions and run tests with them.

**Comment by dicrescenzoa** [comment]
Hi @bnusunny, it would be great if you can take a look into it 👍 

at the moment I've tried this approach but still the adapter seems is not executed.

```Dockerfile
FROM golang:1.22-alpine AS build_base
RUN apk add --no-cache git
WORKDIR /tmp/gin

COPY . .
RUN go mod download
RUN GOOS=linux CGO_ENABLED=0 go build -o main .


FROM alpine:latest
RUN apk add --no-cache bash
COPY --from=public.ecr.aws/awsguru/aws-lambda-adapter:0.8.2 /lambda-adapter /opt/extensions/lambda-adapter
ADD https://github.com/aws/aws-lambda-runtime-interface-emulator/releases/latest/download/aws-lambda-rie /usr/bin/aws-lambda-rie

RUN chmod 755 /usr/bin/aws-lambda-rie

COPY ./entry.sh /
RUN chmod 755 /entry.sh

COPY --from=build_base /tmp/gin/main  /main

ENV PORT=8081 READINESS_CHECK_PORT=8081 READINESS_CHECK_PATH=/ping

ENTRYPOINT ["/entry.sh"]
```

```bash
#!/bin/sh

# Start the first process
./main &
status=$?
if [ $status -ne 0 ]; then
  echo "Failed to start my server: $status"
  exit $status
fi


if [ -z "${AWS_LAMBDA_RUNTIME_API}" ]; then
  exec /usr/bin/aws-lambda-rie "$@"
else
  exec "$@"
fi
```
Plese let me know if you need any help with it!

**Comment by mxfactorial** [comment]
your shell script doesnt start the `main` binary with the emulator

and the emulator expects a [bootstrap](https://github.com/aws/aws-lambda-runtime-interface-emulator/blob/71388dd788b7a5519262391ce73fe6548dbaf86e/cmd/aws-lambda-rie/main.go#L66) file

so [give](https://github.com/awslabs/aws-lambda-web-adapter/blob/main/examples/gin/app/Dockerfile) one

then extensions start in a [sandbox](https://github.com/aws/aws-lambda-runtime-interface-emulator/blob/71388dd788b7a5519262391ce73fe6548dbaf86e/cmd/aws-lambda-rie/main.go#L67-L71)

**Comment by bnusunny** [comment]
Converting to a discussion
